src/third/redux-persist/createPersistor.js

```diff
@@ -57,13 +57,18 @@ export default function createPersistor (store, config) {
         writes.push([key, serializer(substate)])
       }
 
-      // Warning: not guaranteed to be done in a transaction.
-      storage.multiSet(
-        writes.map(([key, serializedSubstate]) => [createStorageKey(key), serializedSubstate])
-      ).catch(warnIfSetError(updatedSubstates));
+      try {
+        // Warning: not guaranteed to be done in a transaction.
+        await storage.multiSet(
+          writes.map(([key, serializedSubstate]) => [createStorageKey(key), serializedSubstate])
+        )
+      } catch (e) {
+        warnIfSetError(updatedSubstates)(e)
+        throw e
+      }
 
       writeInProgress = false
-      lastState = state
+      lastWrittenState = state
```

Comment:
As mentioned above, this update should happen atomically with where the state is compared with the old value. That'll make this new name not so accurate.

I agree it'd be good to make the name more specific like this. I'm not immediately thinking of an equally pithy name, though -- it's basically the last state we've *started* writing.

… Hmmm. I'm now thinking about the possibility that the write can fail, and realizing that that points back in the direction of the "last written" semantics that this branch gives it -- effectively that means that when a write fails, the next time we have an update we'll retry writing the keys that we tried to write last time, as well as any that have newly changed. (Though it isn't perfect at that: if some `state.foo` changes from A to B, we have an error writing that update, and then it changes back to A, we won't try to write A for it, even though it's possible that the update to B actually made it to disk. This mostly… hmm, I was going to say this doesn't end up mattering for us because our subtrees are objects that get new identities every time, but it can actually matter where we use Immutable collections, because e.g. a new empty `Immutable.Map` is identical to any previous empty `Immutable.Map`. … Thinking some more, I think to properly handle this it's probably best to do so explicitly, with like a `Set` of keys that we've started writes to and aren't sure we succeeded. I can try doing that after this PR is in -- I don't think it's critical compared to the holes this subsystem inherently has due to the races that this PR mitigates.)

I guess the upshot of that last paragraph is: I think it still makes best sense to do this update atomically with where the state gets compared with the old value. That makes it clear what this variable's job is, and it makes it easy to reason through at least what the sequence of *attempted* writes is: it's the diffs from one state, to another, to another, etc., with the endpoint of one attempted write being the starting point of the next, and `_resetLastState` effectively suppresses one link in that chain from getting written out. The chain of *actual* writes gets a gap if a link fails, but at least it's straightforward to see that, and we'll fix it separately.

Reply:
Aha, and I see that both the name `lastWrittenState`, and its getting updated here at the end, are as in [my draft](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/data.20storage/near/1214393). Heh.

Well, it does have this hole I mentioned above:
> if some `state.foo` changes from A to B, we have an error writing that update, and then it changes back to A, we won't try to write A for it, even though it's possible that the update to B actually made it to disk.

And probably we should fix that. But as I said:
> I don't think it's critical compared to the holes this subsystem inherently has due to the races that this PR mitigates

and really that applies to anything about the case where a write fails, because I don't think that's likely, and if it does happen I'm not sure we have much chance of not just being doomed. (If it does happen, most likely the filesystem is full or the storage device is failing, and any subsequent write attempt is likely to fail too.)

So I'm happy merging this series with the logic it has in this respect and fixing it later. No need to move the update next to the comparison.